src/main/scala/analysis/Cfg.scala

```diff
@@ -51,382 +196,930 @@ trait CfgNode:
-    val cfg: Cfg = Cfg.generateCfgProgram(prog, false)
-    new IntraproceduralProgramCfg(prog, cfg)
+    val (entryNode: CfgFunctionEntryNode, exitNode: CfgFunctionExitNode) = procToCfg(proc)
+    val (newEntry: CfgFunctionEntryNode, newExit: CfgFunctionExitNode) = (entryNode.copyNode(), exitNode.copyNode())
 
-/** Control-flow graph for a program, where function calls are represented as expressions, without using call/after-call
- * nodes.
- */
-class IntraproceduralProgramCfg(prog: Program, cfg: Cfg) extends ProgramCfg(prog, cfg)
+    callToNodes += (newEntry -> mutable.Set[CfgCommandNode]())
 
+    // Entry is guaranteed to only have one successor (by our cfg design)
+    var currNode: CfgNode = entryNode.succ(intra = true).head
+    visitNode(currNode, newEntry, TrueLiteral)
 
-object InterproceduralProgramCfg:
+    /** Walk this proc's cfg until we reach the exit node on each branch. We do this recursively, tracking the previous
+      * node, to account for branches and loops.
+      *
+      * We can't represent the parameters as an edge as one node comes from the old cfg, and the other from the new cfg.
+      *
+      * @param node
+      *   Node in the original procedure's cfg we're up to cloning
+      * @param prevNewNode
+      *   The originating node in the new clone's cfg
+      * @param cond
+      *   The condition leading to `node` from `prevNewNode`
+      */
+    def visitNode(node: CfgNode, prevNewNode: CfgNode, cond: Expr): Unit = {
 
-  /** Generates an [[InterproceduralProgramCfg]] from a program.
-   */
-  def generateFromProgram(prog: Program): InterproceduralProgramCfg =
-    val cfg: Cfg = Cfg.generateCfgProgram(prog, true)
-    new InterproceduralProgramCfg(prog, cfg)
+      if (node == exitNode) {
+        cfg.addEdge(prevNewNode, newExit, cond)
+        return;
+      }
 
-/** Control-flow graph for a program, where function calls are represented as expressions, without using call/after-call
- * nodes.
- */
-class InterproceduralProgramCfg(prog: Program, cfg: Cfg) extends ProgramCfg(prog, cfg)
+      // Link this node with predecessor in the new cfg
+      val newNode = node.copyNode()
+      cfg.addEdge(prevNewNode, newNode, cond)
 
+      // Update cfg information
+      node match {
+        case n: CfgCommandNode =>
+          n.data match {
+            case d: DirectCall =>
+              procToCalls(proc) += newNode
+                .asInstanceOf[CfgCommandNode] // This procedure (general) is calling another procedure
+              callToNodes(newEntry) += newNode
+                .asInstanceOf[CfgCommandNode] // This procedure (specfic) is calling another procedure
+              procToCallers(d.target) += newNode.asInstanceOf[CfgCommandNode] // Target of this call has a new caller
+            case i: IndirectCall =>
+              procToCalls(proc) += newNode.asInstanceOf[CfgCommandNode]
+              callToNodes(newEntry) += newNode.asInstanceOf[CfgCommandNode]
```

Comment:
Why not type match before copying, instead of copying, type matching the original, and then casting the copy on every use?

Reply:
Resolved. [Upcoming merge from yousif-memory-region-analysis]